### PR TITLE
[PIPE-1069] Add log appender

### PR DESCRIPTION
**Pull request by Pezzah**
## Goal

Add a log appender to the java library so that errors can be sent to Bugsnag via a logger

## Tests

Tests added to cover new log appender functionality, and check the existing functionality works via the appender

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: examples/logback/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Bugsnag Logback Java Example
+
+A Spring Boot example application to show how to use the Bugsnag Logback appender in a Spring based Java application.
+
+- Run the app
+
+    ```shell
+    ../../gradlew clean bootRun
+    ```
```
Comment (fractalwrench) [inline examples/logback/README.md]: We should add instructions on how to change the API key

File: src/main/java/com/bugsnag/Bugsnag.java [hunk 2]
```diff
@@ -42,6 +42,44 @@ public void rejectedExecution(Runnable runnable, ThreadPoolExecutor executor) {
      * Initialize a Bugsnag client and automatically send uncaught exceptions.
      *
      * @param apiKey your Bugsnag API key from your Bugsnag dashboard
+     * @return The Bugsnag client instance
+     */
+    public static Bugsnag createBugsnag(String apiKey) {
+        return createBugsnag(apiKey, true);
```
Comment (fractalwrench) [inline src/main/java/com/bugsnag/Bugsnag.java]: We should be consistent in how we name this across platforms - Android uses `init`, and others use `start`. @kattrali did you have any thoughts on this? The notifier library spec doesn't seem to be too clear on this point

File: src/main/java/com/bugsnag/Bugsnag.java [hunk 3]
```diff
@@ -42,6 +42,44 @@ public void rejectedExecution(Runnable runnable, ThreadPoolExecutor executor) {
...
+
+                // Ensure that sendUncaughtExceptions is respected
+                BugsnagAppender.getInstance().getBugsnag()
+                        .config.setAutoCaptureSessions(sendUncaughtExceptions);
+
+                return BugsnagAppender.getInstance().getBugsnag();
+            }
+        }
+
+        return new Bugsnag(apiKey, sendUncaughtExceptions);
```
Comment (fractalwrench) [inline src/main/java/com/bugsnag/Bugsnag.java]: Some customers initialise Bugsnag multiple times in different parts of their app. I think it would be useful to have some way of disabling the appender on a per-client basis, and generally assuming that the appender will have to co-operate with multiple clients. The `ExceptionHandler` is a good example for how this is handled.

File: src/main/java/com/bugsnag/Bugsnag.java [hunk 4]
```diff
@@ -42,6 +42,44 @@ public void rejectedExecution(Runnable runnable, ThreadPoolExecutor executor) {
...
+     * @return The Bugsnag client instance
+     */
+    public static Bugsnag createBugsnag(String apiKey, boolean sendUncaughtExceptions) {
+
+        // Look for an existing instance of Bugsnag in the appender
+        if (BugsnagAppender.getInstance() != null
+                && BugsnagAppender.getInstance().getBugsnag() != null) {
+
+            // Check that the API key matches
+            if (apiKey.equals(BugsnagAppender.getInstance().getBugsnag().config.apiKey)) {
```
Comment (fractalwrench) [inline src/main/java/com/bugsnag/Bugsnag.java]: We need to validate the API key with a null check as per the constructor

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+        }
+    }
+
+    /**
+     * @return An object mapper to serialize into the logging context with
+     */
+    protected static ObjectMapper getObjectMapper() {
+        if (mapper == null) {
+            mapper = new ObjectMapper();
+        }
```
Comment (fractalwrench) [inline src/main/java/com/bugsnag/BugsnagAppender.java]: I think there's a chance that multiple instances could be created here, as this will be called from multiple threads

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+    /**
+     * @see Bugsnag#setSessionEndpoint(String)
+     */
+    public void setSessionEndpoint(String sessionEndpoint) {
+        this.sessionEndpoint = sessionEndpoint;
+
+        if (bugsnag != null) {
+            bugsnag.setSessionEndpoint(sessionEndpoint);
+        }
+    }
```
Comment (fractalwrench) [inline src/main/java/com/bugsnag/BugsnagAppender.java]: We've recently moved to using `setEndpoints` instead

File: examples/logback/src/main/resources/logback.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<configuration>
+	<include resource="org/springframework/boot/logging/logback/base.xml" />
+
+	<appender name="BUGSNAG" class="com.bugsnag.BugsnagAppender">
+		<apiKey>your-api-key-here</apiKey>
+        <releaseStage>staging</releaseStage>
+        <appVersion>1.0.1</appVersion>
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply examples/logback/src/main/resources/logback.xml]: added

File: build.gradle [hunk 8]
```diff
@@ -28,7 +28,14 @@ dependencies {
     compile 'com.fasterxml.jackson.core:jackson-databind:2.9.1'
     compile 'org.slf4j:slf4j-api:1.7.25'
     compileOnly 'javax.servlet:javax.servlet-api:3.1.0'
+    compileOnly('ch.qos.logback:logback-classic:1.2.3') {
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply build.gradle]: No, are there particular things to consider about versions? It seems to be a bit complicated already with multiple versions being detected

File: examples/logback/README.md (hunk 1 above)
Comment (Pezzah) [reply examples/logback/README.md]: added

File: src/main/java/com/bugsnag/Bugsnag.java (hunk 2 above)
Comment (Pezzah) [reply src/main/java/com/bugsnag/Bugsnag.java]: changed to `init`. Makes sense to be consistent with Android

File: src/main/java/com/bugsnag/Bugsnag.java (hunk 3 above)
Comment (Pezzah) [reply src/main/java/com/bugsnag/Bugsnag.java]: I think the most sensible way of handling this would be to have multiple log appenders defined in the XML file.

Added code to handle muliple instances of the log appender defined

File: src/main/java/com/bugsnag/Bugsnag.java (hunk 4 above)
Comment (Pezzah) [reply src/main/java/com/bugsnag/Bugsnag.java]: added

File: src/main/java/com/bugsnag/BugsnagAppender.java (hunk 5 above)
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: Changed to initialize at the top of the file

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 9]
```diff
@@ -0,0 +1,681 @@
...
+
+    /**
+     * Adds the given key / value to the current thread logging context
+     * to be used for the next report
+     *
+     * @param key the key to add
+     * @param value the value to add
+     */
+    public static void addReportMetaData(String tab, String key, Object value) {
```
Comment (simonbowring) [inline src/main/java/com/bugsnag/BugsnagAppender.java]: We should clarify that the metadata will only be added to the next report, then it gets removed.

File: examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        } catch (RuntimeException e) {
+            LOGGER.info(e.getMessage(), e);
+        }
+
+        // Send a handled exception with custom MetaData
+        LOGGER.info("Sending a handled exception to Bugsnag with custom MetaData");
+        try {
+            throw new RuntimeException("Handled exception - custom metadata");
+        } catch (RuntimeException e) {
+            BugsnagAppender.addReportMetaData("report", "something", "that happened");
```
Comment (simonbowring) [inline examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java]: It's not clear from these example strings that the args are for tab, key and value.

File: src/test/java/com/bugsnag/AppenderTest.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Test for the Bugsnag Appender
+ * NOTE: Not called BugsnagAppenderTest because that throws away errors to prevent cycles
+ */
+public class AppenderTest {
```
Comment (simonbowring) [inline src/test/java/com/bugsnag/AppenderTest.java]: I agree with @fractalwrench, this is a little light on test scenarios.

File: examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+        // Add meta data that will be added to all reports on the current thread
+        BugsnagAppender.addThreadMetaData("report", "application", "has started on this thread");
+
+        // Send a handled exception to Bugsnag
+        LOGGER.info("Sending a handled exception to Bugsnag");
+        try {
+            throw new RuntimeException("Handled exception - default severity");
+        } catch (RuntimeException e) {
+            LOGGER.warn(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java]: changed

File: src/test/java/com/bugsnag/AppenderTest.java [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+        // Check the meta data is set as expected
+        // Should have both report and thread meta data
+        Notification notification = delivery.getNotifications().get(0);
+        assertTrue(notification.getEvents().get(0).getMetaData().containsKey("report"));
+        assertTrue(notification.getEvents().get(0).getMetaData().containsKey("thread"));
+
+        // Should have just thread meta data
+        notification = delivery.getNotifications().get(1);
+        assertFalse(notification.getEvents().get(0).getMetaData().containsKey("report"));
+        assertTrue(notification.getEvents().get(0).getMetaData().containsKey("thread"));
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/test/java/com/bugsnag/AppenderTest.java]: added checks

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 14]
```diff
@@ -0,0 +1,681 @@
...
+    private static ObjectMapper mapper = new ObjectMapper();
+
+    /** Bugsnag API key; the appender doesn't do anything if it's not available. */
+    private String apiKey;
+
+    /** Whether or not to send unhandled exceptions to Bugsnag */
+    private Boolean sendUncaughtExceptions = true;
+
+    /** Whether or not to automatically capture session information */
+    private Boolean autoCaptureSessions = true;
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: changed

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 15]
```diff
@@ -0,0 +1,681 @@
...
+    private List<String> filteredProperties = new ArrayList<String>();
+
+    /** Exception classes to be ignored. */
+    private List<String> ignoredClasses = new ArrayList<String>();
+
+    /** Release stages that should be notified. */
+    private List<String> notifyReleaseStages = new ArrayList<String>();
+
+    /** Project packages. */
+    private List<String> projectPackages = new ArrayList<String>();
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: changed to Set

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 16]
```diff
@@ -0,0 +1,681 @@
...
+        }
+    }
+
+    @Override
+    public void start() {
+        if (apiKey != null && !apiKey.isEmpty()) {
+            this.bugsnag = createBugsnag();
+
+            instances.put(apiKey, this);
+        }
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: Attempting to initialize Bugsnag with no API key will throw a null pointer exception at the moment. Do we need to duplicate the logic here, or were you expecting different behaviour?

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 17]
```diff
@@ -0,0 +1,681 @@
...
+    /**
+     * Checks to see if a stack trace came from the Bugsnag library
+     * (prevent possible infinite reporting loops)
+     *
+     * @param throwable the exception to check
+     * @return true if the stacktrace contains a frame from the Bugsnag library
+     */
+    private boolean detectLogFromBugsnag(Throwable throwable) {
+
+        // Check all places that LOGGER is called with an exception in the Bugsnag library
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: I'm not sure how better to protect against this, we can't exclude all of com.bugsnag because we wouldn't get any of the logs from our apps. We should avoid logging with an error message in this code.

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 18]
```diff
@@ -0,0 +1,681 @@
...
+        for (StackTraceElement element : throwable.getStackTrace()) {
+            if (element.getClassName().startsWith("com.bugsnag.Bugsnag")
+                    || element.getClassName()
+                    .startsWith("com.bugsnag.delivery.OutputStreamDelivery")
+                    || element.getClassName()
+                    .startsWith("com.bugsnag.delivery.SyncHttpDelivery")) {
+                return true;
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: added a list

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 19]
```diff
@@ -0,0 +1,681 @@
...
+
+    /**
+     * Gets the throwable from the log event (if any)
+     * @param event The log event
+     * @return The throwable (or null)
+     */
+    private Throwable extractThrowable(ILoggingEvent event) {
+        IThrowableProxy throwableProxy = event.getThrowableProxy();
+        if (throwableProxy instanceof ThrowableProxy) {
+            return ((ThrowableProxy) event.getThrowableProxy()).getThrowable();
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: changed

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 20]
```diff
@@ -0,0 +1,681 @@
...
+        if (filteredProperties.size() > 0) {
+            bugsnag.setFilters(filteredProperties.toArray(new String[0]));
+        }
+
+        bugsnag.setIgnoreClasses(ignoredClasses.toArray(new String[0]));
+
+        if (notifyReleaseStages.size() > 0) {
+            bugsnag.setNotifyReleaseStages(notifyReleaseStages.toArray(new String[0]));
+        }
+        bugsnag.setProjectPackages(projectPackages.toArray(new String[0]));
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: Apparrently this is the recommended way now https://stackoverflow.com/questions/174093/toarraynew-myclass0-or-toarraynew-myclassmylist-size

Also intelliJ prompts me to use a zero sized array if I change it

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 21]
```diff
@@ -0,0 +1,681 @@
...
+
+    /**
+     * Deserializes the contents of the value
+     *
+     * @param value The value to deserialize
+     * @return The value as an object
+     */
+    private static Object getObjectValue(String value) {
+        try {
+            return mapper.readValue(value, Map.class);
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: Any object will deserialize as a map, but it could also be an array, so added a check to see if it is an array or not

File: build.gradle [hunk 22]
```diff
@@ -28,9 +28,16 @@ dependencies {
...
+        exclude group: 'org.slf4j'
+    }
+
 
+    testCompile('ch.qos.logback:logback-classic:1.2.3') {
+        exclude group: 'org.slf4j'
+    }
     testCompile 'junit:junit:4.12'
-    testCompile 'org.slf4j:slf4j-simple:1.7.25'
+    testCompile 'org.slf4j:log4j-over-slf4j:1.7.25'
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply build.gradle]: yes, we need to be able to do LOGGER.warn() etc..

File: examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+        LOGGER.info("Sending a handled exception to Bugsnag with custom MetaData");
+        try {
+            throw new RuntimeException("Handled exception - custom metadata");
+        } catch (RuntimeException e) {
+            BugsnagAppender.addReportMetaData("report", "something", "that happened");
+            BugsnagAppender.addReportMetaData("report", "something else", "that happened");
+            LOGGER.warn(e.getMessage(), e);
+        }
+
+        // Test an unhanded exception from a different thread as shutdown hooks
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java]: fixed

File: examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+                throw new RuntimeException("Unhandled exception");
+            }
+        };
+
+        thread.start();
+
+        // Wait for unhandled exception thread to finish before exiting
+        thread.join();
+
+        // Remove the thread meta data so it won't be added to future reports
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java]: added

File: examples/logback/src/main/resources/application.properties [hunk 25]
```diff
@@ -0,0 +1 @@
+logging.level.com.bugsnag=DEBUG
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply examples/logback/src/main/resources/application.properties]: I don't think we need this file

File: src/main/java/com/bugsnag/BugsnagAppender.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: TIL 👍

File: src/main/java/com/bugsnag/Bugsnag.java [hunk 26]
```diff
@@ -42,6 +42,39 @@ public void rejectedExecution(Runnable runnable, ThreadPoolExecutor executor) {
...
+    public static Bugsnag init(String apiKey, boolean sendUncaughtExceptions) {
+        if (apiKey == null) {
+            throw new NullPointerException("You must provide a Bugsnag API key");
+        }
+
+        // Look for an existing instance of Bugsnag in the appender
+        BugsnagAppender appender = BugsnagAppender.getInstance(apiKey);
+
+        if (appender != null && appender.getBugsnag() != null) {
+            return appender.getBugsnag();
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/Bugsnag.java]: I don't think that we can perform the setup again (or remove the setup) if `sendUncaughtExceptions` is different

File: src/main/java/com/bugsnag/Bugsnag.java [hunk 27]
```diff
@@ -128,6 +162,17 @@ public Delivery getDelivery() {
         return config.delivery;
     }
 
+    /**
+     * Get the delivery to use to send sessions.
+     *
+     * @return the delivery to use to send sessions.
+     * @see Delivery
+     */
+    public Delivery getSessionDelivery() {
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/Bugsnag.java]: Added for completeness (and might be needed for testing). The equivalent method exists for the notify endpoint

File: src/main/java/com/bugsnag/BugsnagAppender.java (hunk 9 above)
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: added note

File: examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java (hunk 10 above)
Comment (Pezzah) [reply examples/logback/src/main/java/com/bugsnag/example/logback/cli/ApplicationCommandLineRunner.java]: changed names

File: src/test/java/com/bugsnag/AppenderTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/test/java/com/bugsnag/AppenderTest.java]: added more tests

Comment (Pezzah) [reply src/test/java/com/bugsnag/AppenderTest.java]: added more tests

File: src/main/java/com/bugsnag/logback/MetaData.java [hunk 28]
```diff
@@ -0,0 +1,35 @@
+package com.bugsnag.logback;
+
+/** Used to allow meta data to be added in the logback.xml file */
+public class MetaData {
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/logback/MetaData.java]: renamed to `LogbackMetaData` and changed the internal structure to make it clearer

File: src/main/java/com/bugsnag/logback/MetaData.java [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+
+    private String value;
+
+    public String getTabName() {
+        return tabName;
+    }
+
+    public void setTabName(String tabName) {
+        this.tabName = tabName;
+    }
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/logback/MetaData.java]: added comments

File: examples/logback/src/main/resources/logback.xml [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+        <!--<notifyReleaseStage>production</notifyReleaseStage>-->
+        <!--<notifyReleaseStage>development</notifyReleaseStage>-->
+
+        <!--<projectPackage>com.company.package1</projectPackage>-->
+        <!--<projectPackage>com.company.package2</projectPackage>-->
+
+        <!--<metaData>-->
+            <!--<tabName>myTab</tabName>-->
+            <!--<key>myKey</key>-->
+            <!--<value>myValue</value>-->
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply examples/logback/src/main/resources/logback.xml]: changed the internal structure to make it clearer

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 31]
```diff
@@ -0,0 +1,720 @@
...
+            bugsnag.setNotifyReleaseStages(notifyReleaseStages.toArray(new String[0]));
+        }
+        bugsnag.setProjectPackages(projectPackages.toArray(new String[0]));
+        bugsnag.setSendThreads(sendThreads);
+
+        // Add a callback to put global meta data on every report
+        bugsnag.addCallback(new Callback() {
+            @Override
+            public void beforeNotify(Report report) {
+                for (MetaData metaData : globalMetaData) {
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: changed to a simpler loop

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 32]
```diff
@@ -0,0 +1,720 @@
...
+public class BugsnagAppender extends UnsynchronizedAppenderBase<ILoggingEvent> {
+
+    private static final String LOGGING_CONTEXT_THREAD_PREFIX
+            = "com.bugsnag.BugsnagAppender.thread.";
+    private static final String LOGGING_CONTEXT_REPORT_PREFIX
+            = "com.bugsnag.BugsnagAppender.report.";
+    private static final String LOGGING_CONTEXT_TAB_SEPARATOR = ".reportTab.";
+
+    /** Classes that we should not send logs for (to prevent infinite loops on error) */
+    private static final List<String> EXCLUDED_CLASSES = new ArrayList<String>() {{
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: changed

File: src/test/java/com/bugsnag/AppenderTest.java [hunk 33]
```diff
@@ -0,0 +1,604 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Test for the Bugsnag Appender
+ * NOTE: Not called BugsnagAppenderTest because that throws away errors to prevent cycles
+ */
+public class AppenderTest {
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/test/java/com/bugsnag/AppenderTest.java]: extracted delivery stubs into files, and extracted meta data tests

File: src/main/java/com/bugsnag/BugsnagAppender.java [hunk 34]
```diff
@@ -0,0 +1,720 @@
...
+    public void setSessionDelivery(Delivery delivery) {
+        if (bugsnag != null) {
+            bugsnag.setSessionDelivery(delivery);
+        }
+    }
+
+    // Setters
+
+    /**
+     * Should only be used via the logback.xml file
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply src/main/java/com/bugsnag/BugsnagAppender.java]: added

**Comment by simonbowring** [comment]
I'm going to merge this down into the `spring-improvements` branch, then raise a new PR when that branch is ready
